### PR TITLE
Add support for http retries

### LINKED ISSUE #1012: Feature: Support HTTPS Retries
#### Issue type:

- :heavy_plus_sign: Feature request 

____
#### Description:

In order to better support unstable internet connections, Comunica should have a context entry option to enable retries on failed network calls. I would say that the number and frequency of retries should be configurable.

### PR DESCRIPTION
**Pull request by Laurin-W**
Allow setting httpRetryCount and httpRetryDelay to ActorHttpFetch context.

Closes #1012

### CODE REVIEW

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts [hunk 1]
```diff
@@ -76,12 +76,46 @@ export class ActorHttpFetch extends ActorHttp {
...
+            }
+          }
+        }
+        // If retry count was 0, i.e. only one error was recorded, just throw it.
+        if (errors.length === 1) {
+          throw errors[0];
+        } else {
+          throw new Error(`Number of fetch retries exceeded. Encountered errors: ["${errors.join('", "')}"]`);
+        }
+      };
```
Comment (jeswr) [inline packages/actor-http-fetch/lib/ActorHttpFetch.ts:115]: Rather than reporting all the errors I would just report the first error encountered. I think it is fairly safe to assume that the same problem is going to be encountered each time and reporting all the errors is going to create massive error messages if you do a significant number of retries.

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts [hunk 2]
```diff
@@ -443,5 +443,67 @@ This error can be disabled by modifying the 'httpBodyTimeout' and/or 'httpTimeou
...
+
+    it('should retry with a delay', async() => {
+      let numberOfRetries = 0;
+      const customFetch = jest.fn(async(_, init) => {
+        if (numberOfRetries < 2) {
+          numberOfRetries++;
+          throw new Error('Retry count not reached.');
+        }
+        return {};
+      });
```
Comment (jeswr) [inline packages/actor-http-fetch/test/ActorHttpFetch-test.ts:455]: Rather than having this logic you could probably just test how many times the fetch function was called https://stackoverflow.com/questions/52885551/jest-check-how-many-times-a-mocked-module-function-was-called

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts [hunk 3]
```diff
@@ -443,5 +443,67 @@ This error can be disabled by modifying the 'httpBodyTimeout' and/or 'httpTimeou
...
+        context: new ActionContext({
+          [KeysHttp.fetch.name]: customFetch,
+          [KeysHttp.httpRetryCount.name]: 2,
+          [KeysHttp.httpRetryDelay.name]: 100,
+        }),
+      });
+
+      const time = performance.now() - start;
+      expect(time).toBeGreaterThan(200);
+      expect(time).toBeLessThan(250);
```
Comment (jeswr) [inline packages/actor-http-fetch/test/ActorHttpFetch-test.ts:470]: What guarantees it being lest than `250`?

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts (hunk 2 above)
Comment (jeswr) [reply packages/actor-http-fetch/test/ActorHttpFetch-test.ts:455]: Ah I see you have it below - why is this needed as well?

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts [hunk 4]
```diff
@@ -76,12 +76,46 @@ export class ActorHttpFetch extends ActorHttp {
...
+            if (requestInit.signal?.aborted) {
+              throw error;
+            }
+            errors.push(error);
+            // Wait for specified delay.
+            if (retryCount >= 0) {
+              await new Promise((resolve, reject) => {
+                // Cancel waiting, if timeout is reached.
+                requestInit.signal?.addEventListener('abort', () => {
+                  reject(new Error('Fetch aborted by timeout.'));
```
Comment (jeswr) [inline packages/actor-http-fetch/lib/ActorHttpFetch.ts:102]:
```suggestion
                requestInit.signal?.on('abort', reason => {
                  reject(reason ?? new Error('Request aborted before timeout.'));
```

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts [hunk 5]
```diff
@@ -76,12 +76,46 @@ export class ActorHttpFetch extends ActorHttp {
...
+          try {
+            return await (customFetch || fetch)(action.input, requestInit);
+          } catch (error: unknown) {
+            // If the fetch was aborted by timeout, we won't retry.
+            if (requestInit.signal?.aborted) {
+              throw error;
+            }
+            errors.push(error);
+            // Wait for specified delay.
+            if (retryCount >= 0) {
```
Comment (jeswr) [inline packages/actor-http-fetch/lib/ActorHttpFetch.ts:98]: Won't this always be true?

**Review by jeswr — CHANGES_REQUESTED** [review]
Nice work! I've left some passing comments.

I haven't looked at the aborting behavior in a while - in particular the behavior of the `fetchInitPreprocessor` so I will leave someone who has looked into that more recently to check the abort logic; otherwise I will have a proper look tommorow.

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts (hunk 3 above)
Comment (Laurin-W) [reply packages/actor-http-fetch/test/ActorHttpFetch-test.ts]: The rest of the test does not take more than 10 ms at max (basically just calls the mockedFetch). So I thought, it might be reasonable to expect the whole test to be within that bounds and larger time spans indicate that something is going wrong.

Comment (jeswr) [reply packages/actor-http-fetch/test/ActorHttpFetch-test.ts]: I think leave it out just because the tests suites are run concurrently and so is a little unpredictable and has the potential to become quite flakey as a test and cause workflow failures.

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts (hunk 2 above)
Comment (Laurin-W) [reply packages/actor-http-fetch/test/ActorHttpFetch-test.ts:455]: Not sure if I'm getting you right here. The point of counting the retries was that I want the fetch function to throw twice but resolve on the third time. The page you linked just talks about testing the number of function calls.

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 1 above)
Comment (rubensworks) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts:115]: I agree with the above.
We could optionally append something like `(failed for x retries)` to the error message.

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts [hunk 6]
```diff
@@ -76,12 +76,46 @@ export class ActorHttpFetch extends ActorHttp {
...
       requestInit = await this.fetchInitPreprocessor.handle(requestInit);
+      let retryCount: number = action.context?.get(KeysHttp.httpRetryCount) ?? 0;
 
-      // Perform request
       const customFetch: ((input: RequestInfo, init?: RequestInit) => Promise<Response>) | undefined = action
         .context?.get(KeysHttp.fetch);
 
-      const response = await (customFetch || fetch)(action.input, requestInit);
+      // Actual fetch respecting timeout and retry count.
+      const getResponse = async(): Promise<Response> => {
```
Comment (rubensworks) [inline packages/actor-http-fetch/lib/ActorHttpFetch.ts:85]: The `run` method is starting to become quite long now.

Shall we move this inline function to a dedicated (possibly static) method on `ActorHttpFetch`?

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts [hunk 7]
```diff
@@ -76,12 +76,46 @@ export class ActorHttpFetch extends ActorHttp {
...
+        // When retry count is greater than 0, record the errors for failed requests.
+        const errors: unknown[] = [];
+        while (retryCount-- >= 0) {
+          try {
```
Comment (rubensworks) [inline packages/actor-http-fetch/lib/ActorHttpFetch.ts:89]: Open question: next to catching network errors, should we also handle server errors (probably only those in the 500 range then)?

Comment (rubensworks) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts:89]: By default, probably not, but an option for this might be valuable.

File: packages/context-entries/lib/Keys.ts [hunk 8]
```diff
@@ -48,6 +48,15 @@ export const KeysHttp = {
    * but until the response is fully consumed.
    */
   httpBodyTimeout: new ActionContextKey<boolean>('@comunica/bus-http:http-body-timeout'),
+  /**
+   * Number of retries to make on failed network calls. This only takes effect
+   * on errors thrown during the initial fetch() call and not while streaming the body.
```
Comment (rubensworks) [inline packages/context-entries/lib/Keys.ts:53]: > and not while streaming the body.

I get why. But perhaps it should?
What do you think @jeswr?

Complexity of this PR would increase quite a bit with this though, so if we do this, we may want to do it in a separate PR.

File: packages/context-entries/lib/Keys.ts [hunk 9]
```diff
@@ -48,6 +48,15 @@ export const KeysHttp = {
...
+   */
+  httpRetryCount: new ActionContextKey<number>('@comunica/bus-http:http-retry-count'),
+  /**
+   * Delay in milliseconds to wait between fetch retries.
+   */
+  httpRetryDelay: new ActionContextKey<number>('@comunica/bus-http:http-retry-delay'),
```
Comment (rubensworks) [inline packages/context-entries/lib/Keys.ts:59]: Might be good to also mention that this requires `httpRetryCount` to be set.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Really nice stuff @Laurin-W!

Just some minor nits and questions.

File: packages/context-entries/lib/Keys.ts (hunk 8 above)
Comment (jeswr) [reply packages/context-entries/lib/Keys.ts:53]: I agree we should tackle it at some point but I also think it should be a separate PR.


In order to do it properly we would have to add code to find the point in the retried stream at which the original stream broke and continue from there (since Comunica will have already started processing the first stream which broke; and we don't want to introduce broken results).

It would also be completely different error handling implementation wise because rather than throwing an error in the actor as we do here; we would instead need to be doing error handling on the stream.

Comment (rubensworks) [reply packages/context-entries/lib/Keys.ts:53]: Yeah, it will not be trivial, that's for sure.
There's probably even a research paper in there :-)

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 4 above)
Comment (Laurin-W) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts:106]: Regarding the `numberOfRetries` variable of the first comment. That's for passing the request on the third attempt. Or is there a jest-ier way of keeping track of this?

Regarding the second comment with the `signal?.on(...))` proposal - that does not seem to work as `signal` only exposes a lambda called `onabort` which is however undefined in that context.

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 5 above)
Comment (Laurin-W) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts:100]: Not on the last allowed try in which case the loop is subsequently exited and and the encountered error is thrown.

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 7 above)
Comment (Laurin-W) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts]: added an option httpRetryOn5xx :)

**Comment by Laurin-W** [comment]
Thanks for the feedback!
I made the change suggested and changed a couple of variable names and comments to be more explicit, I hope :)

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 4 above)
Comment (jeswr) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts]: > Regarding the numberOfRetries variable of the first commen

Oh - that makes sense! Apologies for the slightly rushed review on my end last night :)

> Regarding the second comment with the signal?.on(...)) proposal

Ok - no worries; I'd assumed this was extending the `EventEmitter` object but I guess not.

File: packages/actor-http-fetch/test/ActorHttpFetch-test.ts (hunk 2 above)
Comment (jeswr) [reply packages/actor-http-fetch/test/ActorHttpFetch-test.ts]: Yep - I was just rushed when I did the review and missed the intention. This looks good!

File: packages/actor-init-query/lib/cli/CliArgsHandlerBase.ts [hunk 10]
```diff
@@ -123,6 +123,18 @@ export class CliArgsHandlerBase implements ICliArgsHandler {
...
         },
+        httpRetryCount: {
+          type: 'number',
+          describe: 'The number of retries to perform on failed fetch requests',
+        },
+        httpRetryDelay: {
+          type: 'number',
+          describe: 'The number of milliseconds to wait between fetch retries',
+        },
+        httpRetryOn5xx: {
```
Comment (jeswr) [inline packages/actor-init-query/lib/cli/CliArgsHandlerBase.ts:134]: I'm wondering if this would be better named as `httpRetryOnServerError`. Probably wouldn't bother changing it unless other changes need be made as well.

**Review by jeswr — APPROVED** [review]
Fantastic work! Just left a minor comment around naming.

File: packages/actor-http-fetch/lib/ActorHttpFetch.ts (hunk 7 above)
Comment (rubensworks) [reply packages/actor-http-fetch/lib/ActorHttpFetch.ts]: Nice!

File: packages/actor-init-query/lib/cli/CliArgsHandlerBase.ts (hunk 10 above)
Comment (rubensworks) [reply packages/actor-init-query/lib/cli/CliArgsHandlerBase.ts:134]: Had this in mind as well, so might be good to change still indeed.

Comment (Laurin-W) [reply packages/actor-init-query/lib/cli/CliArgsHandlerBase.ts]: Got that changed :+1:

**Comment by Laurin-W** [comment]
From my side, this is good to go then :)

**Comment by rubensworks** [comment]
Really nice work @Laurin-W, thanks a lot!
